### PR TITLE
Fix issue https://github.com/dotnet/efcore/issues/21844

### PR DESCRIPTION
**Pull request by mikewodarczyk**
EFCore now skips constructors generated with c#9 that contain parameters with no names.  
The nameless constructor parameters were throwing null reference exceptions when
EFCore was trying to modify the strings to the various conventions for matching
properties like pascal-case, camel-case, m_, etc.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/efcore?pullRequest=21846) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/efcore?pullRequest=21846) before we can accept your contribution.<br/><br/>:x: mikewodarczyk [sign now](https://cla.dotnetfoundation.org/dotnet/efcore?pullRequest=21846)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/efcore?pullRequest=21846) it.</sub>

File: test/EFCore.Record.Tests/IdAndStringRecord.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+
+#nullable enable
+
+namespace EFCore.Record.Tests
+{
+    public record IdAndStringRecord
```
Comment (AndriySvyryd) [inline test/EFCore.Record.Tests/IdAndStringRecord.cs:12]: Move this to `EFCore.Specification.Tests\WithConstructorsTestBase` and remove the new project

Comment (mikewodarczyk) [reply test/EFCore.Record.Tests/IdAndStringRecord.cs:12]: @AndriySvyryd The "record" tag used in this test is new to c# 9 preview which only compiles under a TargetFramework of "net5.0".  The EFCore.Specification.Tests is built under "netcoreapp3.1".   If I move my test, as you request, I would have to change the target framework of that project.  

Alternatively, I could try and write a test under netcoreapp3.1 that mimics the c# 9 compiler by creating a constructor with a nameless parameter.  

Is there a separate place where "net5.0" specific tests will eventually go that I could use? or will the test code-base move to net5.0 at some time in the future? 

What is your recommendation?

Comment (smitpatel) [reply test/EFCore.Record.Tests/IdAndStringRecord.cs:12]: Test projects are cross compiled on netcoreapp5.0 & netcoreapp3.1 tfms. You should be able to use ifdef to write a test which skips on netcoreapp3.1 if the feature is not available there.

Comment (mikewodarczyk) [reply test/EFCore.Record.Tests/IdAndStringRecord.cs:12]: @smitpatel.  Thanks!!  Doing a conditional compile on #if NET5_0 gets me half way there.  I also need the language version to be 9.0 (or preview) for the c# 9 record syntax to be recognized.  It does not look like the language version sets any symbols that I can conditionally compile on.  Even if it did, I don't think that EFCore is running a multi-target compile under the c# 9 (or preview) language version so my test wouldn't get run anyway.

Comment (AndriySvyryd) [reply test/EFCore.Record.Tests/IdAndStringRecord.cs:12]: @mikewodarczyk You can add `<LangVersion>9</LangVersion>` to EFCore.Specification.Tests.csproj

File: All.sln [hunk 2]
```diff
@@ -102,11 +102,11 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "EFCore.SqlServer.Benchmarks
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "EFCore.Sqlite.Benchmarks", "benchmark\EFCore.Sqlite.Benchmarks\EFCore.Sqlite.Benchmarks.csproj", "{5A44F67E-517F-458D-B9C7-12A4DBBAD64A}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "EFCore.Benchmarks", "benchmark\EFCore.Benchmarks\EFCore.Benchmarks.csproj", "{2642F4F0-69BE-4C43-94B7-B298FEC87D89}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "EFCore.Benchmarks", "benchmark\EFCore.Benchmarks\EFCore.Benchmarks.csproj", "{2642F4F0-69BE-4C43-94B7-B298FEC87D89}"
```
Comment (smitpatel) [inline All.sln:105]: Revert these changes.

File: test/EFCore.Record.Tests/EFCore.Record.Tests.csproj [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (smitpatel) [inline test/EFCore.Record.Tests/EFCore.Record.Tests.csproj:1]: Remove Record.Tests

File: test/EFCore.Specification.Tests/ImmutableModelTestBase.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore.TestModels.ImmutableRecordModel;
+using Xunit;
+
+namespace Microsoft.EntityFrameworkCore
+{
+#if NET5_0
+    class ImmutableModelTestBase<TFixture> : IClassFixture<TFixture>
```
Comment (smitpatel) [inline test/EFCore.Specification.Tests/ImmutableModelTestBase.cs:15]: avoid new test class. Add test to `WithConstructorsTestBase` with appropriate class/test added to that context with ifdef

File: test/EFCore.Specification.Tests/WithConstructorsTestBase.cs [hunk 5]
```diff
@@ -738,6 +738,49 @@ public virtual void Query_with_loader_delegate_injected_into_property_via_constr
...
+            var immutableBlog = new BlogAsImmutableRecord(aBlogTitle);
+
+            context.Add(immutableBlog);
+            await context.SaveChangesAsync();
+
+            Assert.NotEqual(0, immutableBlog.BlogId);
+            Assert.Equal(aBlogTitle, immutableBlog.Title);
+        }
+#endif
+
```
Comment (smitpatel) [inline test/EFCore.Specification.Tests/WithConstructorsTestBase.cs:756]:
```suggestion
```

File: test/EFCore.Specification.Tests/WithConstructorsTestBase.cs [hunk 6]
```diff
@@ -738,6 +738,49 @@ public virtual void Query_with_loader_delegate_injected_into_property_via_constr
             Assert.Same(blog, blog.LazyPcsPosts.Skip(1).First().LazyPcsBlog);
         }
 
+#if NET5_0
+        [ConditionalFact]
+        public virtual async Task Add_immutable_record()
+        {
+            using var context = CreateContext();
+            const string aBlogTitle = "xyzzy";
```
Comment (smitpatel) [inline test/EFCore.Specification.Tests/WithConstructorsTestBase.cs:746]:
```suggestion
            var blogTitle = "xyzzy";
```

File: test/EFCore.Specification.Tests/WithConstructorsTestBase.cs [hunk 7]
```diff
@@ -1539,6 +1582,16 @@ public abstract class WithConstructorsFixtureBase : SharedStoreFixtureBase<WithC
 
             protected override void OnModelCreating(ModelBuilder modelBuilder, DbContext context)
             {
+
+#if NET5_0
+                modelBuilder.Entity<BlogAsImmutableRecord>(
+                   b =>
+                   {
+                       b.HasKey("_blogId");
```
Comment (smitpatel) [inline test/EFCore.Specification.Tests/WithConstructorsTestBase.cs:1590]: Where does `_blogId` defined?

**Comment by mikewodarczyk** [comment]
All of the suggested changes are complete.  I forced a revert and made the changes through a new check-in on the main branch.  I have resolved-conversation on all of the changes to mark them as complete.

**Comment by smitpatel** [comment]
@mikewodarczyk - There are still changes to add Records.Tests which needs to be removed.

**Comment by smitpatel** [comment]
@mikewodarczyk - Thank you for the contribution.
